RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,823 @@
+
+#### Nullable reference type declarations in F\#
+
+There will be no way to declare an F# type as follows:
+
+```fsharp
+// This will not be possible!
+type (SomeClass | null)() = class end
+```
+
+And in existing F# today, the only way to declare an F# reference type that could be `null` in F# code is to use `[<AllowNullLiteral>]`. This would not change with this proposal; that is, the only way to declare a reference type in F# that could have `null` as a value in F# code is `[<AllowNullLiteral>]`.
+
+### Checking of nullable references
+
+There are a few common ways that F# programmers check for null today. Unless explicitly mentioned as unsupported, support for flow analysis is under consideration for certain null-checking patterns.
+
+#### Likely supported: Pattern matching with alias
+
+```fsharp
+let len (str: string | null) =
+    match str with
+    | null -> -1
+    | s -> s.Length // OK - we know 's' is string
+```
+
+This is by far the most common pattern in F# programming. In the previous code sample, `str` is of type `string | null`, but `s` is now of type `string`. This is because we know that based on the `null` has been accounted for by the `null` pattern.
+
+#### Likely supported: isNull and others using null-handling decoration
+
+```fsharp
+let len (str: string | null) =
+    if isNull str then
+        -1
+    else
+        str.Length // OK
+```
+
+The `isNull` function is considered to be a good way to check for `null`. We can annotate it in FSharp.Core with the attribute used for functions handling null (we'll call it `[<HandlesNull>]`).
+
+Similarly, CoreFX will annotate methods like `String.IsNullOrEmpty` as handling `null`. This means that the most common of null-checking methods can be respected by flow analysis and reduce the amount of nullability assertions.
+
+##### Possibly supported: Pattern matching with wildcard
+
+```fsharp
+let len (str: string | null) =
+    match str with
+    | null -> -1
+    | _ -> str.Length // OK - 'str' is string
+```
+
+A slight riff that is less common is treating `str` differently when in a wildcard pattern's scope. To reach that code path, `str` must indeed by non-null, so this could potentially be supported.
+
+#### Possibly supported: If check
+
+```fsharp
+let len (str: string | null) =
+    if str = null then
+        -1
+    else
+        str.Length // OK - 'str' is a string
+```
+
+Although less common, this is a valid way to check for `null` and could be enabled by the compiler. We know for certain that `str` is a `string` in the `else` branch. The inverse check is also true.
+
+#### Possibly supported: After null check within boolean expression
+
+```fsharp
+let len (str: string | null) =
+    if (str <> null && str.Length > 0) then // OK - `str` must be null
```

Comment:
I think C# is gonna steel the idea from Kotlin as is https://kotlinlang.org/docs/reference/null-safety.html#checking-for-null-in-conditions :) I wrote a bit of Kotlin some time ago and I must say, it works really well there, but still feels a bit hacky, like a work around some problem (no real pattern matching probably?)

Reply:
The issues I have with not supporting this are:

* Failing to do this will result in warnings on perfectly `null`-safe code that users will expect the compiler _not_ to yell at them about
* For cases like the example shown, it _appears_ simple for a compiler to support (even though it's not simple at all), which to me is a signal that it should be considered
* Expectations may already be in place for this kind of support, even if the languages people use as examples are not in the ML family (though this is something people may come to accept)

I agree that it's not a particularly in the F# style of things, but a success metric of this feature is having very, very, very low false negatives. If we emit them _by design_, that will just be frustrating for people.